File: src/org/elixir_lang/credo/Service.java
package org.elixir_lang.credo;

import com.intellij.openapi.components.*;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@com.intellij.openapi.components.State(
    name = "Credo",
    storages = {
        @Storage(file = StoragePathMacros.PROJECT_FILE),
        @Storage(file = StoragePathMacros.PROJECT_CONFIG_DIR + "/" + "credo.xml", scheme = StorageScheme.DIRECTORY_BASED)
    }
)
public class Service implements PersistentStateComponent<State> {
    @NotNull
    private State state = new State();

    @NotNull
    public static Service getInstance(@NotNull Project project) {
        Service persisted = ServiceManager.getService(project, Service.class);
        Service service;

        if (persisted != null) {
            service = persisted;
        } else {
            service = new Service();
        }

        return service;
    }

    @Nullable
    @Override
    public State getState() {
        return state;
    }

    @Override
    public void loadState(State state) {
        this.state = state;
    }

    public boolean includeExplanation() {
        return state.includeExplanation;
    }

    public void includeExplanation(boolean includeExplanation) {
        state.includeExplanation = includeExplanation;
    }
}

File: src/org/elixir_lang/credo/Service.kt
package org.elixir_lang.credo

import com.intellij.execution.configuration.EnvironmentVariablesData
import com.intellij.openapi.components.PersistentStateComponent
import com.intellij.openapi.components.Storage
import com.intellij.openapi.project.Project

@com.intellij.openapi.components.State(name = "Credo", storages = [Storage(value = "credo.xml")])
class Service : PersistentStateComponent<State?> {
    private var state = State()
    override fun getState(): State = state

    override fun loadState(state: State) {
        this.state = state
    }

    var elixirArguments: String
        get() = state.elixirArguments
        set(value) {
            state.elixirArguments = value
        }

    var erlArguments: String
        get() = state.erlArguments
        set(value) {
            state.erlArguments = value
        }

    var environmentVariableData: EnvironmentVariablesData
        get() = state.environmentVariableData.toConfiguration()
        set(value) {
            state.environmentVariableData =
                org.elixir_lang.credo.state.EnvironmentVariablesData.fromConfiguration(value)
        }

    companion object {
        @JvmStatic
        fun getInstance(project: Project): Service = project.getService(Service::class.java) ?: Service()
    }
}
